src/rt/minfo.d

```diff
@@ -53,6 +53,128 @@ struct ModuleGroup
         return _modules;
     }
 
+    // this function initializes the bookeeping necessary to create the
+    // cycle path, and then creates it. It is a precondition that src and
+    // target modules are involved in a cycle.
+    //
+    // The delegate is a helper to map module info pointers to index into the modules array
+    private int[] genCyclePath(int srcidx, int targetidx,
+            scope int delegate(immutable(ModuleInfo)*) findModule)
+    {
+        import core.bitop : bt, btc, bts;
+
+        // set up all the arrays. Use the GC, we are going to exit anyway.
+        int[] cyclePath;
+        int[] distance;
+        int[][] edges;
+        distance.length = _modules.length;
+        edges.length = _modules.length;
+        immutable nwords = (_modules.length + 8 * size_t.sizeof - 1) / (8 * size_t.sizeof);
+        immutable flagbytes = nwords * size_t.sizeof;
+        auto reachable = cast(size_t*) malloc(flagbytes);
+        scope (exit)
+            .free(reachable);
+
+        foreach (i, m; _modules)
+        {
+            // use bit array to prevent duplicates
+            // https://issues.dlang.org/show_bug.cgi?id=16208
+            memset(reachable, 0, flagbytes);
+            foreach (e; m.importedModules)
+            {
+                auto impidx = findModule(e);
+                if (impidx != -1 && impidx != i)
+                    if (!bts(reachable, impidx))
+                        edges[i] ~= impidx;
+            }
+        }
+
+        // determine the shortest path between two modules. Uses dijkstra
+        // without a priority queue. (we can be a bit slow here, in order to
+        // get a better printout).
+        void shortest(int start, int target)
```

Comment:
All the edges, the cycle array that records the path, and the distance array are referenced from the enclosing function. But the distance array is only used locally, so it could potentially be allocated for each run.

In any case, I'm not sure it's too important. Nothing else will use this function, as it's very specific to the cycle detection.